freqtrade/exchange/kraken.py

```diff
@@ -78,6 +78,7 @@ def get_balances(self) -> CcxtBalances:
                     # x["side"], x["amount"],
                 )
                 for x in orders
+                if x["price"] is not None
```

Comment:
well what's the status of that order?

even a conditional order has a price (either a trigger price, or both a trigger and a limit price) - maybe not in the "price" field - but in some other field - so i guess excluding that from the balance calculation of "tied up capital" is wrong - or am i missing something ?

Reply:
Well so the order with the price None was an order which will close the position but the acctual capital that is now in that position is still being showed as another  open trade. The problem was simply caused by me setting an trailing stop loss sell order after a buys order successfull execution. It simply got listed as an open conditional trade with no price and it also should not count towards used capital as the original buy order (when it hasn't yet been executed) should be the "owner" of that capital which it still is.

Its a bit late so sorry to whoever has to read and understand this.

Reply:
Ok, after rethinking and taking into account what you said, it could be that the bot on my side was fine because he simply looked at the amount of EUR left as that's my stake currency. Because I have only one API return that I took for debugging which shows actually that i kind of misinterpreted it a bit as there are two prices, one 0.00000 and the second one None. And after looking into the Kraken API it turns out that the first price the 0.0000 is for the field of "primary price" and as the order hasn't yet executed its 0.00000 while the limit and spot price fields are filled out (as well as the vol field). But the second price key is actually "Average price (quote currency)" which the Kraken API returned as None, while the "Stop price (quote currency)" and "limit price  (quote currency)" are filled correctly with EUR values.

Here is the link to the docs page: https://docs.kraken.com/api/docs/rest-api/get-open-orders

Reply:
I think the approach of mixing manual orders with freqtrade trades is a very dangerous one - as freqtrade will not be aware about that order - will simply close the trade by placing a new order - having your order initiate a reverse (short) position.

---

ccxt will parse "average"  once the order filled - or started to fill. before that it'll be limit and stop_price (or stop_loss_price?) - until the order actually triggers.

Now my concern here is two-fold - which is why i need some time to think about this and "play" with kraken a bit.

It's crashing if price is None - that's pretty clear and obvious - but only if that order was actually a sell order (otherwise price is not used).

On the other hand - if we assume this shouldn't include orders that "don't tie up capital" (it's not used capital?) - then the assumption that this is "only" orders without a price might fail - and the actual determination should be based on order-type (maybe additionally).

I'm also a bit surprised that trailing-stop doesn't tie up capital to be honest - as in my testing (given, it's been a while) - regular stoploss orders do tie up capital - effectively preventing the placement of non-stop orders until that order is closed.

Reply:
> I think the approach of mixing manual orders with freqtrade trades is a very dangerous one - as freqtrade will not be aware about that order - will simply close the trade by placing a new order - having your order initiate a reverse (short) position.

Yes, but the thing is that it still crashed even if I blacklisted the pair, so the bot crashes regardless of if It's supposed to watch the Pair or not.
Also, additionally I thought that even if the bot trades on that pair it would kind of ignore the trade as it has its own trade database (trade not made by bot so simply ignore), I've also noticed that it would do as you said.


> It's crashing if price is None - that's pretty clear and obvious - but only if that order was actually a sell order (otherwise price is not used).

But then why is the price None order crashing it even if it's a `side: "buy",` / `type: "buy",` and therefore its price field shouldn't be used? Cause after rethinking everything The stop loss buy order alone did manage to crash the bot with it price None the OTO I set didn't affect it and isn't even included in the data that is being parsed

Reply:
can't tell without seeing the actual response you got from ccxt at that time if i'm honest.

it's possible that ccxt parsed the side wrongly (assumign stop order = sell) - but it's speculation without seeing the actual response object.

Reply:
Here is one response object I got

[
  {
    id: "[REDACTED]",
    clientOrderId: "0",
    info: {
      id: "[REDACTED]",
      refid: None,
      userref: "0",
      status: "open",
      opentm: "1725814282.1196225",
      starttm: "0",
      expiretm: "0",
      descr: {
        pair: "XBTEUR",
        type: "buy",
        ordertype: "trailing-stop",
        price: "+4.7000%",
        price2: "0",
        leverage: "5:1",
        order: "buy 0.01990500 XBTEUR @ trailing stop +4.7000%",
        close: "close position @ trailing stop -6.0000%",
      },
      vol: "0.01990500",
      vol_exec: "0.00000000",
      cost: "0.00000",
      fee: "0.00000",
      price: "0.00000",
      stopprice: "50917.80000",
      limitprice: "48632.10000",
      misc: "",
      oflags: "fciq",
      margin: True,
      trigger: "index",
    },
    timestamp: 1725814282119,
    datetime: "2024-09-08T16:51:22.119Z",
    lastTradeTimestamp: None,
    status: "open",
    symbol: "BTC/EUR",
    type: "trailing-stop",
    timeInForce: None,
    postOnly: False,
    side: "buy",
    price: None,
    stopPrice: 50917.8,
    triggerPrice: 50917.8,
    takeProfitPrice: None,
    stopLossPrice: None,
    cost: None,
    amount: 0.019905,
    filled: 0.0,
    average: None,
    remaining: 0.019905,
    reduceOnly: None,
    fee: { cost: "0.00000", rate: None, currency: "EUR" },
    trades: [],
    fees: [{ cost: 0.0, rate: None, currency: "EUR" }],
    lastUpdateTimestamp: None,
  },
];

Reply:
I think the new approach will work - it's explicitly excluding non-sell orders without price (Or if remaining is None for whatever unknown reason).

balance calls shouldn't really fail in any case - as they can include orders not directly tied to the bot - so whatever random combi you can have on the exchange.